native/cocos/renderer/gfx-metal/MTLSemaphore.h

```diff
@@ -32,38 +32,52 @@ namespace gfx {
 
 class CCMTLSemaphore final {
 public:
-    explicit CCMTLSemaphore(uint32_t initialValue) : _semaphoreCount(initialValue) {
+    explicit CCMTLSemaphore(uint32_t initialValue) : _semaphoreCount(initialValue), _restCount(initialValue) {
         _semaphore = dispatch_semaphore_create(initialValue);
     }
-    ~CCMTLSemaphore() = default;
+    
+    ~CCMTLSemaphore() {
+        if(_semaphore) {
+            while(_restCount-- >0) {
+                dispatch_semaphore_signal(_semaphore);
+            }
+            dispatch_release(_semaphore);
+            _semaphore = nullptr;
+        }
+    };
     CCMTLSemaphore(const CCMTLSemaphore &) = delete;
     CCMTLSemaphore(CCMTLSemaphore &&) = delete;
     CCMTLSemaphore &operator=(const CCMTLSemaphore &) = delete;
     CCMTLSemaphore &operator=(CCMTLSemaphore &&) = delete;
 
-    void signal() const {
+    void signal() {
         dispatch_semaphore_signal(_semaphore);
+        --_restCount;
```

Comment:
OperatIng `int32_t _restCount;` in different thread is dangerous since it isn't atomic.